server/services/applicationService.test.ts

```diff
@@ -155,6 +155,84 @@ describe('ApplicationService', () => {
     })
   })
 
+  describe('appendToList', () => {
+    const application = applicationFactory.build({ data: null })
+    const token = 'some-token'
+    const request = createMock<Request>({
+      params: { id: application.id, task: 'some-task', page: 'some-page' },
+      user: { token },
+    })
+    const applicationData = createMock<UpdateCas2Application>()
+
+    beforeEach(() => {
+      applicationClient.find.mockResolvedValue(application)
+    })
+
+    describe('when there are no validation errors', () => {
+      let page: DeepMocked<TaskListPage>
+
+      beforeEach(() => {
+        page = createMock<TaskListPage>({
+          errors: () => {
+            return {} as TaskListErrors<TaskListPage>
+          },
+          body: { foo: 'bar' },
+        })
+        ;(getPageName as jest.Mock).mockReturnValue('some-page')
+        ;(getTaskName as jest.Mock).mockReturnValue('some-task')
+      })
+
+      it('does not throw an error', () => {
+        expect(async () => {
+          await service.appendToList(page, request)
+        }).not.toThrow(ValidationError)
+      })
+
+      it('saves data to the api', async () => {
+        await service.appendToList(page, request)
+
+        expect(applicationClientFactory).toHaveBeenCalledWith(token)
+        expect(applicationClient.update).toHaveBeenCalledWith(application.id, applicationData)
+      })
+
+      it('updates an in-progress application', async () => {
+        application.data = { 'some-task': { 'some-page': [{ question: 'answer' }] } }
+        const requestWithBody = createMock<Request>({
+          body: { newQuestion: 'new answer' },
+          params: { id: application.id, task: 'some-task', page: 'some-page' },
+          user: { token },
+        })
+
+        await service.appendToList(page, requestWithBody)
+
+        expect(getApplicationUpdateData).toHaveBeenCalledWith({
+          ...application,
+          data: {
+            'some-task': {
+              'some-page': [{ question: 'answer' }, { newQuestion: 'new answer' }],
+            },
+          },
+        })
```

Comment:
not really related to this PR, but if we're sending the whole application to the API every time we save, is it really a PUT request?

Reply:
which method were you thinking might be more appropriate?

In general, I would say that PUT is appropriate for any form of update. 

However, some teams / frameworks prefer to use PATCH for a partial update as distinct from a full replacement, which I think is technically more correct. Perhaps we should use a PATCH here? But I would say that change could follow on as an enhancement.

A couple of references:

- [PUT vs PATCH & PUT vs POST](https://dev.to/mehmehmehlol/put-vs-patch-put-vs-post-56i9)
- [Using PATCH in express](https://www.tabnine.com/code/javascript/functions/express/Express/patch)

Reply:
I think it just feels weird to me because we're passing around the whole JSON blob, so in a sense it feels more like a POST. But I guess that's just how it is when you have this data type? Whatever the operation (unless it's DELETE), you run your filtering logic in the UI, and then send the whole thing to the API...?